src/openzaak/utils/admin.py

```diff
@@ -223,9 +227,75 @@ def delete_model(self, request, obj):
 
         super().delete_model(request, obj)
 
-        self.trail(obj, request, action, data, None)
+        self.trail(obj, viewset, request, action, data, None)
 
     def delete_queryset(self, request, queryset):
         # data before
         for obj in queryset:
             self.delete_model(request, obj)
+
+    def save_formset(self, request, form, formset, change):
+        """
+        Given an inline formset save it to the database.
+        """
+        if not hasattr(formset, "viewset"):
+            super().save_formset(request, form, formset, change)
+            return
+
+        viewset = formset.viewset
+
+        # we need to save data before update/delete
+        obj_before_data = {}
+        for form in formset.initial_forms:
+            obj = form.instance
+            if obj.pk is None:
+                continue
+
+            obj_before = obj.__class__.objects.get(pk=obj.pk)
+            data = self.get_serializer_data(request, viewset, obj_before)
+            obj_before_data.update({obj.uuid: data})
+
+        super().save_formset(request, form, formset, change)
+
+        # delete existing
+        for obj in formset.deleted_objects:
+            data_before = obj_before_data[obj.uuid]
+
+            self.trail(
+                obj, viewset, request, CommonResourceAction.destroy, data_before, None
+            )
+
+        # change existing
+        for obj, changed_data in formset.changed_objects:
+            data_before = obj_before_data[obj.uuid]
+            data_after = self.get_serializer_data(request, viewset, obj)
+
+            self.trail(
+                obj,
+                viewset,
+                request,
+                CommonResourceAction.update,
+                data_before,
+                data_after,
+            )
+
+        # add new
+        for obj in formset.new_objects:
+            data_after = self.get_serializer_data(request, viewset, obj)
+            self.trail(
+                obj, viewset, request, CommonResourceAction.create, None, data_after
+            )
+
+
+class AuditTrailInlineAdminMixin(object):
+    viewset = None
```

Comment:
Maybe we should add a check if viewset is still None on init and raise an ImproperlyConfigured exception to prevent accidental left outs.

Reply:
In this case we couldn't trail changes in objects with no related viewset. Like EnkelvoudigInformatieObjectCanonical